src/loki/search.cc

```diff
@@ -658,8 +658,23 @@ struct bin_handler_t {
         correlated.edges.erase(new_end, correlated.edges.end());
       }
 
+
+      // Store filtered edges for retry in case find path with correlated.edges failed.
+
+      // We will also add penalty to the score of filtered edges, so as to handle the
+      // case low score filtered_edges VS high score good heading edges during retry.
+      // We use max score of the correlated.edgesfor penalty.
+      float penalty = 0;
+      for (const auto &edge : correlated.edges) {
+        penalty = std::max(edge.score, penalty);
+      }
+      for (auto &filtered_edge : filtered) {
+        filtered_edge.score = filtered_edge.score + penalty;
+        correlated.filtered_edges.push_back(std::move(filtered_edge));
+      }
+
```

Comment:
Sounds good to me for now. Ultimately we want to update edge scores to incorporate more factors. It could be that edge scores could be based on difference of heading from the input heading in such a way that this could work on the first pass.